### PR TITLE
Add env vars to set default kubectl context and ns

### LINKED ISSUE #27308: Proposal: make `kubectl` respect KUBECTL_{CONTEXT,NAMESPACE} environment variables
I'm curious if there is interest in creating some environment variables (`KUBECTL_CONTEXT` and `KUBECTL_NAMESPACE`) to make `kubectl` a bit easier to use when one has multiple clusters configured in `~/.kube/config`.

The alternative to these environment variables either require a lot of typing (specify `--context` and `--namespace` to all `kubectl` commands). Or as I've done so far, wrapper functions in my shell, but I've been unable to setup my environment such that I can have helper functions without breaking the tab completion for `kubectl`.

Is there interest in this? Or asked differently, would such a PR be accepted?

### PR DESCRIPTION
**Pull request by awh**
* KUBECTL_CONTEXT sets the default for --context
* KUBECTL_NAMESPACE sets the default for --namespace

Fixes kubernetes/kubernetes#27308.

**Release note**:
```release-note
`kubectl` now observes `KUBECTL_CONTEXT` and `KUBECTL_NAMESPACE` to set defaults for `--context` and `--namespace` respectively.
```

### CODE REVIEW

**Comment by awh** [comment]
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify.

Please recheck my CLA registration @k8s-ci-robot :+1:

**Comment by simonswine** [comment]
I am not too sure if this will break anything. But the functionality would be really valuable IMHO

/ok-to-test

**Comment by awh** [comment]
The `pull-kubernetes-bazel-build` and `pull-kubernetes-bazel-test` jobs have been in SUCCESS state for days but GH hasn't realised - perhaps running them again will help:

/test pull-kubernetes-bazel-build
/test pull-kubernetes-bazel-test

**Comment by awh** [comment]
@simonswine any chance of getting this into 1.10 do you think?

**Comment by liggitt** [comment]
I don't think there is consensus that this is a good direction to go in, and the deadline for 1.10 changes has passed. 

/hold

**Comment by soltysh** [comment]
I'll second Jordan. I'm worried that having these env variables increases the risk you will destroy/break a namespace you're not actually wanted to modify. I've done stuff like that in the past where mistakenly I've run a destructive command on a production environment instead of a testing one. Having an environment variable being used instead of explicitly requiring a namespace|context can do more harm than good.

**Comment by Quentin-M** [comment]
This does borderline sound like nightmare.. The whole thing (flags, merging of the different files in `$KUBECONFIG`, `.kube/config`...) is complicated enough in my opinion. 

As of today, to gain a fair confidence you are not going to screw up, you must `kubectl config view`, parse the manifests you are working for a `metadata.namespace` field, and follow your terminal cursor to make sure you don't have a `--namespace` flag around. With this PR, I would need a fourth step and check my environment variable.

Now, you can be pretty sure for yourself and your CI/CD pipeline.. But what about all the other users of your cluster, that use casually `kubectl`, to explore the running resources, debug, and run maintenance/migration steps?  As explained by @soltysh, the risk of executing commands in a context that is not the expected one, is already omnipresent, and that would only make it worse.

Why not use shell aliases, or `kubectl config set-context`?

**Comment by ahmetb** [comment]
I think this is not complicating the code, although adds one more step (and therefore confusion in terms of precedence) to see where the value is coming from. 

Similarly, quite a few `kubectl` environment variables are duplicated in client libraries, such as https://github.com/kubernetes/client-go/ (`KUBECONFIG` is an example, and it works exactly the same way in kubectl and client-go). In this case these env vars probably don't apply to client libraries.

Also, how do you expect people to discover these environment variables? (Is it going to be documented only in the `kubectl --help`? For example KUBECONFIG is documented at https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/).

Disclaimer: I am the author of a fairly popular project in this space https://github.com/ahmetb/kubectx . I think this PR still adds a lot of value, most people want ability to have different contexts/namespaces in different shell tabs.

**Comment by endorama** [comment]
**TL;DR**: this adds a lot of value for custom workflows and do not increase the risk of executing against a wrong context/namespace. 

I'm using context below, but referring to context/namespace.

> I'm worried that having these env variables increases the risk you will destroy/break a namespace you're not actually wanted to modify. 

AFAIU the only way to make sure you are running against the wanted context is to specify it as a flag, and this implementation does not change that ( hope not being wrong on this ).

> I've done stuff like that in the past where mistakenly I've run a destructive command on a production environment instead of a testing one. 

Probably all of us did it at least once and I share your concern on this.

> Having an environment variable being used instead of explicitly requiring a namespace|context can do more harm than good.

Setting the value in the *kubeconfig* has the same risk of setting an environment variable.

> As of today, to gain a fair confidence you are not going to screw up, you must kubectl config view, parse the manifests you are working for a metadata.namespace field, and follow your terminal cursor to make sure you don't have a --namespace flag around. With this PR, I would need a fourth step and check my environment variable.

I'm fairly new to `kubectl`, so I can be missing something, but to check the current context I'm executing command into shouldn't I use `kubectl config current-context`? As a user this seems the only place I should check, and I generally rely on its value. ( if I correctly understood the *you* in your statement )

> Now, you can be pretty sure for yourself and your CI/CD pipeline.. But what about all the other users of your cluster, that use casually kubectl, to explore the running resources, debug, and run maintenance/migration steps? As explained by @soltysh, the risk of executing commands in a context that is not the expected one, is already omnipresent, and that would only make it worse.

The risk is omnipresent and will probably always be. Using aliases do not lessen the risk ( and can probably make it worst due to need of sharing and syncing aliases between developers ).  

Using `kubectl config set-context` could be not appliable if you don't want the *kubeconfig* to be changed.

If I'm getting this right, the proposed solution is along the line of this:

```
$ alias kubectl='kubectl --context ${KUBE_CONTEXT}'
$ export KUBE_CONTEXT=prod # or dev, or wathever
$ kubecl delete pods ...
```

I don't see how this could be less error prone than just setting the variable and executing `kubectl`.

In my particular use case we are commiting the *kubeconfig* to a git repo ( without secrets ), and we would like to avoid having a `current-context` already set in it. This is easily doable ( and with *kubeconfig* merge we can ensure that it doesn't get added running `kubectl config set-context` ), but then to run `kubectl` with a context you always have to set the flag `--context`, at each invocation, from cli by and operator and in each script using the command.  

Sorry for the long post and thank you for the hard work on this project :)

**Comment by endorama** [comment]
> With this PR, if I specify --kubeconfig=/path/to/kubeconfig and the kubeconfig contains current-context: foo, I have no idea what context will actually be used unless I ensure the KUBECTL_CONTEXT envvar is unset. That is problematic.

I agree that this is problematic, and if this is the case I evidently misunderstood how setting config values currently works.

To clarify my opinion, **if** the kubeconfig contains an explicit `current-context`, or the flag `--context/--namespace` is used, that should have precedence. But if it doesn't, reading the value from env vars is not more risky that reading it from a file.

Precedence of different configuration sources is pretty common, and generally env vars are overridden by explicit values. The main point is to provide a default, not an override.

**Comment by ahmetb** [comment]
> To clarify my opinion, if the kubeconfig contains an explicit current-context, [...], that should have precedence

I don't agree with this part. ENV should have more precedence than `current-context` property.

For example, GKE get-credentials command will allways update the current-context, but I love the ability to override it with --context or $KUBE_CONTEXT.

**Comment by youngnick** [comment]
I think the biggest problem with delegating everything to a single setting in a single file is that it's a really poor experience if you're interacting with multiple clusters. If you want to check something in multiple clusters in multiple terminals, you can't without a whole lot of bash shenanigans.

I also don't really understand the complexity objection here, @liggitt. I thought that config file is overridden by env var is overridden by command line flag is a common Unix-style pattern. I've certainly used that scheme often enough previously that I was surprised that kubectl *didn't* do it when I first started using it.

I completely get that there's probably plumbing stuff that I'm not aware of playing in here, it's just that, as an end user who works on a bunch of clusters all day every day, I'd rather that the native tooling supporting me, rather than having to build ramshackle bash structures to make it work.

**Comment by liggitt** [comment]
> I also don't really understand the complexity objection here, @liggitt. I thought that config file is overridden by env var is overridden by command line flag is a common Unix-style pattern.

The current approach makes values explicit. The config file can be overridden by an envvar, but only as a coherent piece (can specify a different config file as a whole). 

Letting implicit envvars override portions of explicit config file values makes it extremely difficult to be sure you are using the config you intend to.

A safe script invoked with a kubeconfig file as input would have to do the following:
``` 
unset KUBECTL_CONTEXT
unset KUBECTL_NAMESPACE
unset ...any other random overriding envvars…
kubectl … --kubeconfig=input.kubeconfig
``` 

Even the discussion in this PR reveals the expected precedence order between file/env/flag is not intuitively understood

**Comment by liggitt** [comment]
> If you want to check something in multiple clusters in multiple terminals, you can't without a whole lot of bash shenanigans.

A kubeconfig file per cluster with a unique KUBECONFIG envvar per terminal seems pretty straightforward to me

**Comment by smarterclayton** [comment]
Given feedback and long standing history, I'm going to close this.  If you'd like to move forward for this, it'll need some stronger arguments for and discussion of the points Jordan raised - probably as a KEP, but starting with a discussion in sig-cli.  There's a lot of very strong reasons to not add additional complexity here and Jordans points are 100% in line with historical precedent.

/close

**Comment by autarchprinceps** [comment]
Why wasn't this implemented? Given the clear feedback on everyone speaking for it, and everyone speaking against, it is clear that this is a desired feature. We certainly want to have and use it, because we want to have separate contexts active in different terminal windows, without much fussing around and ugly workarounds.

**Comment by peterromfeldhk** [comment]
man i did hope i can just add some env vars to change behavior of some existing scripts... but now it seems i have to do a lot more work :(

**Comment by bitti** [comment]
> man i did hope i can just add some env vars to change behavior of some existing scripts... but now it seems i have to do a lot more work :(

Look at https://github.com/danielfoehrKn/kubeswitch. Works with a minimum of setup.
